src/ctap/crypto_wrapper.rs

```diff
@@ -62,6 +89,215 @@ pub fn aes256_cbc_decrypt(
+    pub fn new_ecdsa_from_bytes(bytes: &[u8]) -> Option<Self> {
+        if bytes.len() != 32 {
+            return None;
+        }
+        ecdsa::SecKey::from_bytes(array_ref!(bytes, 0, 32)).map(PrivateKey::from)
+    }
+
+    /// Returns the corresponding public key.
+    pub fn get_pub_key(&self) -> CoseKey {
+        match self {
+            PrivateKey::Ecdsa(ecdsa_key) => CoseKey::from(ecdsa_key.genpk()),
+        }
+    }
+
+    /// Returns the encoded signature for a given message.
+    pub fn sign_and_encode(&self, message: &[u8]) -> Vec<u8> {
+        match self {
+            PrivateKey::Ecdsa(ecdsa_key) => ecdsa_key.sign_rfc6979::<Sha256>(message).to_asn1_der(),
+        }
+    }
+
+    /// The associated COSE signature algorithm identifier.
+    pub fn signature_algorithm(&self) -> SignatureAlgorithm {
+        match self {
+            PrivateKey::Ecdsa(_) => SignatureAlgorithm::ES256,
+        }
+    }
+
+    /// Writes the key bytes.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            PrivateKey::Ecdsa(ecdsa_key) => {
+                let mut key_bytes = vec![0u8; 32];
+                ecdsa_key.to_bytes(array_mut_ref!(key_bytes, 0, 32));
+                key_bytes
+            }
+        }
+    }
+}
+
+impl From<PrivateKey> for cbor::Value {
+    fn from(private_key: PrivateKey) -> Self {
+        cbor_array![
+            cbor_int!(private_key.signature_algorithm() as i64),
+            cbor_bytes!(private_key.to_bytes()),
+        ]
+    }
+}
+
+impl TryFrom<cbor::Value> for PrivateKey {
+    type Error = Ctap2StatusCode;
+
+    fn try_from(cbor_value: cbor::Value) -> Result<Self, Ctap2StatusCode> {
+        let mut array = extract_array(cbor_value)?;
+        if array.len() != 2 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR);
+        }
+        let key_bytes = extract_byte_string(array.pop().unwrap())?;
+        match SignatureAlgorithm::try_from(array.pop().unwrap())? {
+            SignatureAlgorithm::ES256 => PrivateKey::new_ecdsa_from_bytes(&key_bytes)
+                .ok_or(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR),
+            _ => Err(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR),
+        }
+    }
+}
+
+impl From<ecdsa::SecKey> for PrivateKey {
+    fn from(ecdsa_key: ecdsa::SecKey) -> Self {
+        PrivateKey::Ecdsa(ecdsa_key)
+    }
+}
+
+/// Encrypts the given private key and relying party ID hash into a credential ID.
+///
+/// Other information, such as a user name, are not stored. Since encrypted credential IDs are
+/// stored server-side, this information is already available (unencrypted).
+///
+/// Also, by limiting ourselves to private key and RP ID hash, we are compatible with U2F for
+/// ECDSA private keys.
+///
+/// This is v1, and we write the following data for ECDSA (algorithm -7):
+/// -  1 byte : version number
+/// - 16 bytes: initialization vector for AES-256,
+/// -  4 bytes: algorithm in big endian encoding,
```

Comment:
Out of curiosity, is the algorithm sensitive information? I'm wondering if it could be part of the version number (the algorithm would be a function of the version number). This would make sense if we want the client to notify when using old algorithms (the idea you mentioned in another comment) modulo the increased privacy concerns of version number identification.

Reply:
The relying party knows the algorithm, since they store the public key to the encrypted private key. It might also leak due to credential ID size changing. So we could decide to make these bytes "public" as well. Since we'd save the extra AES block, it would make logic a bit easier (good for maintenance) and assertions a bit faster.

By the way, the above mentioned advantage is very theoretical, to be honest. For that to happen, the client needs to have special logic for any OpenSK-based authenticators. The credential wrapping does not have any recommended defaults (part of why we design this ourselves).

Is there anything you had in mind when you said "function of the version number"? I guess just store both like we do now, but we can have a ECDSA v2 and a EdDSA v1 and a COOL_CRYPTO v3 as the respective latest versions?

Reply:
I was thinking if algorithm are changed to better or added as alternative. Because if it's the first, then algorithms are ordered within a version, and I'm assuming versions are also changed for the better, so we have a natural order which can just be a number (we might want to use LEB128 i.e. start using 2 bytes when reaching version 128). That number would fully describe the format that follows (including presence of hmac and its algorithm).

But if algorithms are not ordered within version, then yes, a pair sounds good.

For example, with a single number, decryption would look like:

```rust
match version {
    // Legacy versions
    5 => parse_v5()
    6 => parse_v6()
    7 => parse_v7()
    // Current version (handles are created in this format)
    8 => parse_v8()
    // Unsupported versions (either too old, too new, or invalid)
    _ => error()
}
```

Reply:
I see. So if we want to i.e. upgrade AES and keep all algorithms as is, we'd issue a new version per algorithm. I think that should be good enough.

Reply:
So it would be a real scenario that multiple algorithms live simultaneously while none is better than all others? If yes, I think the pair is alright, and moving out of the encrypted area would just be to save some padding.

Reply:
I think that is a possible scenario. The current EdDSA PR is an example, I don't think EdDSA is strictly better than ECDSA. Also new algorithms could be different key lengths of old versions, and you just hope that the server signals support for the longer key, but allow the older version. But per variant of `PrivateKey`, there will be only 1 active version.

I think I still like the version only encoding, WDYT?

Reply:
(The latest commit implements that.)